compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -226,8 +227,14 @@ impl<'tcx> InferCtxtInner<'tcx> {
     #[inline]
     pub fn unwrap_region_constraints(&mut self) -> RegionConstraintCollector<'_, 'tcx> {
         self.region_constraint_storage
-            .as_mut()
-            .expect("region constraints already solved")
+            // Replace region constraints if they've been taken.
+            // This should only happen on an error path, so delay a bug.
+            .get_or_insert_with(|| {
+                ty::tls::with(|tcx| {
+                    tcx.sess.delay_span_bug(DUMMY_SP, "region constraints already solved");
+                });
+                RegionConstraintStorage::default()
+            })
```

Comment:
Maybe we can keep it named `take_` name, but set some flag that the infos have been taken, and delay a bug if we try to take them again or try to borrow them again?

Reply:
Mm, I have no strong opinion here, but I don't like delaying a bug in InferCtxt. I prefer the code there to be independent from the compilation session as much as possible.

Changing `take_region_var_infos` is unfortunate as it would no longer guard against future addition of region constraints, but I think it's the best option here given that it's used only once in borrowck and we don't expect users of InferCtxt to implement their own region resolution.